### PR TITLE
Table cell `x` and `y` fields [More Flexible Tables Pt.2b]

### PR DESCRIPTION
**Pull request by PgBiel**
Adds fields `x` and `y` to `grid.cell` and `table.cell`, in both ways: not only can you access them in show rules to determine how the cell should be displayed based on its coordinates, but you can also specify `x` and `y` manually for a cell to ensure it will be placed at that particular position (however, will raise an error if that would lead to two cells in the same position).
Specifying `x` but not `y` will lead the cell to be placed in the first available row in that column (or create a row if needed). Similarly, specifying `y` but not `x` will lead the cell to be placed in the first available column in that row (or error if no column is available, since columns are fixed).

Added tests and docs with some examples.

Second part of the second task in https://github.com/typst/typst/issues/3001

~~This PR is a **draft PR** as it depends on #3037~~
~~(For a simpler diff, check out https://github.com/PgBiel/typst/compare/table-cells...PgBiel:typst:cell-positioning )~~

## User-facing API

1. You can now have show rules on cells depend on the cell's final `x` (column) and `y` (row) coordinates:

```typ
// Style based on position
#{
  show table.cell: it => {
    if it.y == 0 {
      strong(it)
    } else if it.x == it.y {
      emph(it)
    } else {
      it
    }
  }
  table(
    columns: 3,
    gutter: 3pt,
    [Name], [Age], [Info],
    [John], [52], [Nice],
    [Mary], [50], [Cool],
    [Jake], [49], [Epic]
  )
}
```
image

2. You can now specify where a cell should be placed:

```typ
#grid(
  columns: 2,
  [A], [B],
  grid.cell(x: 1, y: 2)[C], grid.cell(x: 0, y: 2)[D],
  grid.cell(x: 1, y: 1)[E], grid.cell(x: 0, y: 1)[F],
)
```
image

## Implementation details
- In `CellGrid::resolve`, the grid (vector of cells) is now built from scratch, since cells can now be positioned in any way they wish, so the `cells` vector would have to be reordered; thus, we create a `resolved_cells` vector, and iterate over `cells` to determine the final position of each cell, resolve it, and place it in the appropriate position in the vector (three steps).
- When positioning a cell in an index which would be out of bounds for `cells`, `resolved_cells` expands accordingly. But it has to expand with something (e.g. if a cell is placed 900 positions after the last one, what do we fill the empty vector spaces with?). We _could_ expand adding many normal cells with empty bodies, but then those cells would be taking up space unnecessarily - it should be possible to override those "filler" cells later if we want.
   - Therefore, `resolved_cells` starts out as a vector of `Option<Cell>` - it has `None` where a new cell can be placed.
- An absent cell position (`None` in `Vec<Option<Cell>>`) can be replaced by a cell later in `cells` which might want to occupy its space (turning the position to `Some(cell)`). However, if there are any absent cell positions remaining after rebuilding the grid, they are converted to normal cells with empty bodies (necessary so we can synthesize properties on the empty spaces, e.g. fill or align, and so we can apply show rules on them too).
   - This is done by iterating over `resolved_cells` and converting `Option<Cell>` to `Cell` (just unwrap if the cell is Some, otherwise create a new empty cell).
   - Theoretically this is only necessary if `resolved_cells.len() != cells.len()`, since the only way a `None` entry can remain after the grid is built is if `resolved_cells` was actually expanded. This could lead to a potential optimization here, but that would have to be discussed. For now, this works.

## Remaining questions
- [x] **(Solved)** Should cells specified without overriding any coordinate - `x` or `y` - automatically skip cells which are already occupied by cells with arbitrary/custom position? For example,

```typ
#table(
   columns: 3,
   [A], [B], [C],
   table.cell(x: 1, y: 0),
)
```

will error, which makes sense, since `[B]` was already placed in the grid by the time we got to that last cell. But the following also errors; maybe it shouldn't (and we should just skip the arbitrarily-positioned cell)?

```typ
#table(
   columns: 3,
   table.cell(x: 1, y: 0),
   [A], [B], [C],
)
```
   - **Conclusion:** I made it not error (since that wouldn't be very useful). Instead, cells with fully automatic position (`x: auto` and `y: auto`, the default) will just skip non-absent (non-`None`) positions. (Implemented in c376e5776c084ec521de3f5a69ba09c18b0ca709)

## Performance implications

In my unscientific tests using documents containing only large tables, tables took about **70-80% longer to cold compile** with this PR.

Maybe there isn't too much we can do about this, unfortunately...

## TODO

- [x] ~~**TODO:**~~ Fix problem with incomplete (non-`c`-large) rows
- [x] ~~**TODO:**~~ Docs.
- [x] ~~**TODO:**~~ Merge other PR.

### CODE REVIEW

**Comment by Julian24816** [comment]
Quick suggestion: As this branch depends on the changes on branch `table-cells` you could set that branch as this PRs target branch (and don't merge this/keep it as draft). This will:
- allow you to review only the changes added by this branch without the changes on `table-cells`
- once you merge that other PR, the base branch of this PR will automatically be changed back to `main` by GitHub
- if you review/approve this PR now, this review will not become outdated by merging the other PR

**Comment by PgBiel** [comment]
Hey @Julian24816 , thanks for the suggestion. I believe the problem with your idea is that this branch is in my fork, so the PR would have to be moved to my fork... Do you know if there's a way around this?

**Comment by Julian24816** [comment]
> Hey @Julian24816 , thanks for the suggestion. I believe the problem with your idea is that this branch is in my fork, so the PR would have to be moved to my fork... Do you know if there's a way around this?

Oh, the Smartphone App hid this important detail from me 😅 sadly I don't have any experience with this topic in cross fork PRs...

File: crates/typst/src/layout/grid/layout.rs [hunk 1]
```diff
@@ -278,6 +371,83 @@ impl CellGrid {
...
+/// final index of this cell in the vector of resolved cells.
+fn resolve_cell_position(
+    cell_x: Smart<usize>,
+    cell_y: Smart<usize>,
+    resolved_cells: &[Option<Cell>],
+    auto_index: &mut usize,
+    columns: usize,
+) -> HintedStrResult<usize> {
+    // Translates a (x, y) position to the equivalent index in the final cell vector.
+    let cell_index = |x, y| y * columns + x;
```
Comment (PgBiel) [inline crates/typst/src/layout/grid/layout.rs:386]: Should we use `checked_mul` and `checked_add` here? I was going to do so, but was kinda afraid that this could hurt performance more than it should depending on the amount of cells in the table... maybe it's not too bad?

(Same applies to e.g. replacing `i + next_auto_index` below with `i.checked_add(next_auto_index)` or `*auto_index = resolved_index + 1` with `*auto_index = resolved_index.saturating_add(1)` for example)

Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs:386]: Actually, I think it shouldn't be necessary for the auto_index ones, since for that to overflow you'd need `usize::MAX` cells in your table, and you'd probably run out of memory long before that; if not, you'd try to place a cell in an existing position and fail, so whatever. And I think we should optimize for the cases without arbitrary cell positioning (the vast majority of cases), so that decision could make sense.

OTOH we could just accept the performance hit when arbitrary positioning is used, since it's too easy to screw it up (e.g. add several columns to the table and place a cell at a very high `y`, enough to overflow and have it be placed at some random position possibly without error). Not 100% sure if my logic is correct here though.

Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs:386]: Yeah, just checked... you can produce an overflow with

```typ
#grid(
  columns: 3,
  rows: 2em,
  fill: (x, y) => if calc.odd(x + y) { red.lighten(50%) } else { green },
  grid.cell(y: 6148914691236517207)[a],
)
```

leading to

cell appears in the second row

so yeah, we definitely need something here. Just probably not for the auto index case.

Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs]: Okay, there we go. Added some checks (which shouldn't impact the `auto_index` case) and tests. :+1:
This is probably solved then.

(Though maybe I should use `*auto_index = next_auto_index.wrapping_add(1)` in case people are using debug mode or something to ensure it won't panic... but eh, maybe the paranoia isn't worth it...)

Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs]: I spotted another problem: this will cause a capacity overflow on `.extend` inside the grid building loop.

```typ
#grid(
  columns: 3,
  rows: 2em,
  fill: (x, y) => if calc.odd(x + y) { red.lighten(50%) } else { green },
  grid.cell(x: 2, y: 6148914691236517204)[a],
)
```

It passes the cell index checks (the cell indices are indeed valid), but I assume the vector is trying to reserve some more space, but it can't.

I guess we could use `reserve_exact` before the `extend`, but is it even worth it? Reserving more space is useful in general anyway. Maybe we could just let this one slip.

File: crates/typst/src/layout/grid/layout.rs [hunk 2]
```diff
@@ -200,9 +210,9 @@ impl CellGrid {
         Self { cols, rows, cells, has_gutter, is_rtl }
     }
 
-    /// Resolves all cells in the grid before creating it.
-    /// Allows them to keep track of their final properties and adjust their
-    /// fields accordingly.
+    /// Resolves and positions all cells in the grid before creating it.
+    /// Allows them to keep track of their final properties and positions
+    /// adjust their fields accordingly.
```
Comment (laurmaedje) [inline crates/typst/src/layout/grid/layout.rs:215]: something is off in this sentence

File: crates/typst/src/layout/grid/layout.rs [hunk 3]
```diff
@@ -216,38 +226,123 @@ impl CellGrid {
...
+            );
+
+            // Now let's check if the cell's position is valid.
+            if let Some(existing_cell) = resolved_cells.get_mut(resolved_index) {
+                // We are trying to place a cell in an existing position.
+                // Ensure we aren't trying to place a cell where there is
+                // already one.
+                if existing_cell.is_some() {
+                    bail!(
+                        span,
```
Comment (laurmaedje) [inline crates/typst/src/layout/grid/layout.rs:279]: should we really fail at the generic span here or at the span of the second cell?

File: crates/typst/src/layout/grid/layout.rs [hunk 4]
```diff
@@ -216,38 +226,123 @@ impl CellGrid {
...
+                x,
+                y,
+                &fill.resolve(engine, x, y)?,
+                align.resolve(engine, x, y)?,
+                inset,
+                styles,
+            );
+
+            // Now let's check if the cell's position is valid.
+            if let Some(existing_cell) = resolved_cells.get_mut(resolved_index) {
```
Comment (laurmaedje) [inline crates/typst/src/layout/grid/layout.rs:273]: I think we can simplify this case distinction a bit by doing the following:

1. `if resolved_index >= resolved_cells.len() { resolved_cells.resize(resolved_index + 1, None) }`
2. `let slot = &mut resolved_cells[resolved_index];`
3. Check for existing cell
4. Assign cell

resize + assign is perhaps a tiny bit slower than push but on the upside we have a single code path and less potential for off-by-one bugs.

File: crates/typst/src/layout/grid/layout.rs [hunk 5]
```diff
@@ -216,38 +226,123 @@ impl CellGrid {
...
+        }
+
+        // If not all columns in the last row have cells, we will add absent
+        // positions (later converted to empty cells) and complete the row so
+        // that those positions are susceptible to show rules and receive grid
+        // styling.
+        // We apply '% c' twice so that 'cells_remaining' is zero when the last
+        // row is already filled (then 'resolved_cells.len() % c' would be
+        // zero).
+        let cells_remaining = (c - resolved_cells.len() % c) % c;
```
Comment (laurmaedje) [inline crates/typst/src/layout/grid/layout.rs:315]: We could technically even integrate this into the resize above to ensure that we always push full rows.

File: crates/typst/src/layout/grid/layout.rs [hunk 6]
```diff
@@ -216,38 +226,123 @@ impl CellGrid {
...
-            .chain(std::iter::repeat_with(T::default).take(cells_remaining))
+        let mut resolved_cells: Vec<Option<Cell>> = Vec::with_capacity(cell_count);
+        for cell in cells.iter().cloned() {
+            // Let's calculate the cell's final position based on its
+            // requested position.
+            let resolved_index = {
+                let cell_x = cell.x(styles);
+                let cell_y = cell.y(styles);
+                resolve_cell_position(cell_x, cell_y, &resolved_cells, &mut auto_index, c)
+                    .at(span)?
```
Comment (laurmaedje) [inline crates/typst/src/layout/grid/layout.rs:256]: This should probably also use the cell's span.

If cell spans turn out to be detached, we might need to adjust the place where they are constructed and add an appropriate .spanned().

File: crates/typst/src/layout/grid/layout.rs [hunk 7]
```diff
@@ -278,6 +373,114 @@ impl CellGrid {
...
+                    (Smart::Auto, Smart::Auto) => EcoString::default(),
+                    (Smart::Custom(x), Smart::Auto) => eco_format!("at column {x} "),
+                    (Smart::Auto, Smart::Custom(y)) => eco_format!("at row {y} "),
+                    (Smart::Custom(x), Smart::Custom(y)) => {
+                        eco_format!("at column {x}, row {y} ")
+                    }
+                };
+                HintedString::from(eco_format!(
+                    "the position of a cell {position}would be too large"
+                ))
```
Comment (laurmaedje) [inline crates/typst/src/layout/grid/layout.rs:403]: as far as I can tell, this really only happens for usize overflow? it's nice to handle this, but perhaps the message formatting is a bit overkill. if resolve_cell_position(..).at(..) would use the cell span instead of the grid span, then we extra x/y information isn't all that necessary.

File: crates/typst/src/layout/grid/layout.rs [hunk 8]
```diff
@@ -278,6 +373,114 @@ impl CellGrid {
...
+            }
+            cell_index(cell_x, cell_y)
+        }
+        // Cell has only chosen its column, not its row.
+        (Smart::Custom(cell_x), Smart::Auto) => {
+            if cell_x >= columns {
+                return Err(HintedString::from(eco_format!(
+                    "a cell could not be placed at invalid column {cell_x}"
+                )));
+            }
```
Comment (laurmaedje) [inline crates/typst/src/layout/grid/layout.rs:444]: perhaps we could get rid of this duplication by merging the two match branches?

File: crates/typst/src/layout/grid/layout.rs [hunk 9]
```diff
@@ -278,6 +373,114 @@ impl CellGrid {
...
+            // placed after this one (maybe not immediately after).
+            *auto_index = resolved_index + 1;
+
+            Ok(resolved_index)
+        }
+        // Cell has chosen its exact position.
+        (Smart::Custom(cell_x), Smart::Custom(cell_y)) => {
+            if cell_x >= columns {
+                return Err(HintedString::from(eco_format!(
+                    "a cell could not be placed at invalid column {cell_x}"
```
Comment (laurmaedje) [inline crates/typst/src/layout/grid/layout.rs:433]: if the errors at the cell's span, we could the probably drop the `"a "` and just says `"cell could not ..."`.

File: crates/typst/src/layout/grid/layout.rs [hunk 10]
```diff
@@ -278,6 +373,114 @@ impl CellGrid {
...
+        // Cell has only chosen its row, not its column.
+        (Smart::Auto, Smart::Custom(cell_y)) => {
+            // Let's find the first available column in the chosen row.
+            // The first available position after the first cell of the row,
+            // within up to 'columns' positions, will be the first available
+            // column. We chain extra absent positions (None) in case the
+            // vector isn't large enough yet to hold all cells in the row
+            // (in which case we can place the cell at the first out of bounds
+            // position, since it's available).
+            resolved_cells
```
Comment (laurmaedje) [inline crates/typst/src/layout/grid/layout.rs:465]: I feel like we should probably either use loops consistently or iterators in this function. The first and last are iterators, but the middle one a loop. 

This final one here is a bit tricky to understand. Would something simpler like this below work?
```rust
let start = cell_index(0, cell_y)?;
(start..start + columns)
  .find(|i| !matches!(resolved_cells.get(i), Some(_)))
``` 
Generally, I think directly iterating over the candidate indices might sometimes be simpler than using position() and than mapping over an addition of a base index.

File: crates/typst/src/layout/grid/layout.rs (hunk 3 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs:279]: I think it can be fine to use the span of the second cell, I was mostly just afraid of the span being in some totally unrelated location, leading the error to be potentially confusing. (Could be worse if e.g. you're doing something like `..cells.map(grid.cell)`)

Maybe we shouldn't care that much?

File: crates/typst/src/layout/grid/layout.rs (hunk 2 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs:215]: Git rebase moment :joy:
I'll fix, thanks

File: crates/typst/src/layout/grid/layout.rs (hunk 3 above)
Comment (laurmaedje) [reply crates/typst/src/layout/grid/layout.rs:279]: You can add the table's span to the backtrace with `trace`.

File: crates/typst/src/layout/grid/layout.rs (hunk 5 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs:315]: I considered doing this, but I thought that it doesn't matter that much since we are about to allocate new space for a vector anyway, so if we allocate extra above, it would be a bit of a waste. I also considered that allocating extra space in the loop could be better in terms of performance (as in, we would expect that, if a cell is placed in a certain row, then we'd expect the row to be filled later), but:
1. We already allocate enough to fit the initial cell count at the start, so that'd be a no-op in most cases;
2. Rust is already good at allocating extra space when we don't know how much the vector will expand (which is precisely our case here when arbitrary positioning is used).

Either way, there's no way to confirm that without benchmarks :p. Additionally, one argument in favor of resizing in the loop is that we could assume the invariant that the vector always has full rows, and thus (in `resolve_cell_position`) not have to consider the case where the next available position is out of bounds, other than for cells which request a particular column (then the next available spot in the column could be in a row that doesn't exist in the vector yet).

So we should probably ponder about this a bit. Perhaps it's better for the PR as a whole if we do this change...

File: crates/typst/src/layout/grid/layout.rs (hunk 10 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs:465]: I used to use exclusively loops before, but I thought that it didn't make much sense to do something like `.get(0)`, `.get(1)` etc. when that's precisely iterating over the vector after a certain point (through `.skip`) and calculating the position. (It was especially jarring in the case with `auto_index`, where I'd just add 1 on every iteration, which felt like a code smell :p)  And Rust's standard library probably does this job better than us.

Regarding the final iterator, I had basically that setup you propose, but then I changed to the current one at e10200102a6e478fbed370ce78f190eb42c94d8d due to the reasons above, and it basically works by iterating over the whole row. We can do that since the vector is in row-major order. It's something like `&vec[first_index..(first_index + columns)].iter().position(...)` which will give you the first available column, except that I can't use a slice cuz `first_index + columns` might be out of bounds. If we push full rows, we could use slices, but AFAIK iterators are still faster (at least this [link](https://stackoverflow.com/questions/32690678/is-it-more-efficient-to-slice-an-array-or-use-iteratorskip) suggested so based on some ad-hoc and totally unscientific research).

But the middle one, as far as I'm aware, has to be done with loops since I don't think we can tell iterators to skip `c` elements on each iteration in a very practical way. I considered using `chunks()`, but it didn't feel much better than just using the loop like that.

Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs:465]: With that said, we can probably experiment with returning to loop-based approaches and seeing how they perform.

File: crates/typst/src/layout/grid/layout.rs (hunk 4 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs:273]: I based the current design on the principle of having the common use case (no positioning) be as fast as possible, but perhaps this could make sense if combined with the idea of pushing full rows. Will experiment a bit.

File: crates/typst/src/layout/grid/layout.rs (hunk 10 above)
Comment (laurmaedje) [reply crates/typst/src/layout/grid/layout.rs:465]: My guess would be that this code will be far from being the bottleneck of a Typst execution, so iterator vs range + get will probably have no observable difference.

File: crates/typst/src/layout/grid/layout.rs (hunk 4 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs]: Seems to have mostly equivalent performance, so should be fine. Implemented.

File: crates/typst/src/layout/grid/layout.rs (hunk 10 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs]: Yeah fair enough. Though the `auto_index` case would still worry me. Either way, I switched back to loops now and performance seems to be roughly the same as before, so we should be fine.

File: crates/typst/src/layout/grid/layout.rs (hunk 5 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs]: Okay, well, I did it. In the end I was kinda neutral about it, and there were no performance implications apparently, so I guess it's ok to do it, no hard feelings though.

File: crates/typst/src/layout/grid/layout.rs (hunk 3 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs]: Is using `trace` necessary if the error already occurs from within the table's `Layout`? I was thinking here that, in order to use `trace`, `CellGrid::resolve` might have to take the `Tracing` as a parameter (alongside the `Span`) since it depends on the callee's context, though I guess that wouldn't be that bad.

File: crates/typst/src/layout/grid/layout.rs (hunk 6 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs]: I've implemented an initial version of this. For now I just made it use the table's span when the cell's span is detached. How do you suggest I go about this?

> If cell spans turn out to be detached, we might need to adjust the place where they are constructed and add an appropriate .spanned().

How could this be achieved? Do you mean going through all places in the standard library which build a `GridCell` or `TableCell` and adding `.spanned()` accordingly? Like e.g. in bibliography?

File: crates/typst/src/layout/grid/layout.rs (hunk 8 above)
Comment (laurmaedje) [reply crates/typst/src/layout/grid/layout.rs]: Please mark things that you've implemented as resolved. This helps me see whether the PR is ready for another round of review / merging when looking at it.

**Comment by laurmaedje** [comment]
> I was told otherwise when contributing elsewhere (that the reviewer is responsible for marking threads as resolved)

Interesting. Not sure what the industry practice is, but most contributors here just mark it themselves and it works well enough. Often, I'll notice anyway that something didn't change when doing another review round. But having the marking definitely helps to decide whether another review round should be done.

> However, when I'm unsure about something, I'll probably keep it open and ask explicitly for feedback, otherwise I'll just resolve it. 👍

Sounds good.

File: crates/typst/src/layout/grid/layout.rs (hunk 6 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs]: Okay, I changed bibliography.rs, as far as I could search this is the only place where we create `GridCell` internally, so I made them use the bibliography element's span.

File: crates/typst/src/layout/grid/layout.rs (hunk 3 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs]: Okay, I added a `trace` call to grid and table in the meantime, producing the results below. We can change later if we implement a more general solution in the future.

a `table.cell`-specific error now shows "help: error occurred in this call of function `table`".

a `grid.cell`-specific error now shows "help: error occurred in this call of function `grid`".

File: crates/typst/src/layout/grid/layout.rs (hunk 1 above)
Comment (PgBiel) [reply crates/typst/src/layout/grid/layout.rs]: I think the current overflow checks should be enough to at least avoid unexpected behavior. Crashes are fine imo. Marking this as resolved.

**Comment by laurmaedje** [comment]
Aside from the one conflict, this looks ready to merge.

**Comment by PgBiel** [comment]
> Aside from the one conflict, this looks ready to merge.

I was taking a look, and this conflict seems more significant than I thought; after #3200, we'll probably have to migrate Grid and Table to use `Packed<GridCell>` and `Packed<TableCell>` in order to be able to get each cell's span. But I'm kinda unsure regarding which trait impls should go on GridCell/TableCell directly and which should go on `Packed<GridCell/TableCell>`, considering that `Packed<T>` seems to implement some traits by itself as well (such as `Clone`, though that one is probably fine to use by default). I'm thinking that I'll probably have to implement `ResolvableCell` on `Packed<...Cell>` instead so we can get a cell's span...
...or maybe we could just force `CellGrid` to take a vector of `Packed<T>` where `T: ResolvableCell`. Not sure.

I'll take a further look at this later today.

**Comment by PgBiel** [comment]
Okay, well, I'm not very familiar with the new stuff from #3200, but here's what I did:

1. Initially, in 47d4108cb316234d32e58b1d009aea58d0d4334c, I experimented with having `CellGrid::resolve` take a vector of `Packed<T>` directly. It worked and was fairly simple, but it did introduce some perhaps unnecessary coupling with `Packed`, so I set off to change that.
2. In 5c6b9b46cc52c2e5970c86f04a9a7c0887dc4657, I managed to return `CellGrid::resolve` to its previous code unchanged by implementing `ResolvableCell` and `Default` on `Packed<GridCell/TableCell>`. However, it did require one or two workarounds due to borrow checking stuff (in particular for the `self.push_align(...)` stuff since `self.align()` was being used, leading to concurrent mut / non-mut borrows due to the implicit `deref`/`deref_mut`), but right now I think I prefer this approach so as to keep `CellGrid::resolve` as generic and simplified as possible.

Feel free to give opinions; I can revert 5c6b9b46cc52c2e5970c86f04a9a7c0887dc4657 if that'd be preferred.

**Comment by PgBiel** [comment]
Q: If we keep my current approach of implementing `ResolvableCell` on `Packed<GridCell/TableCell>` (at 4b37f741a05b3c6490c4ca96b1cb1cdf55248f45), is it fine to mutate the `Packed` directly like I'm doing, or should I always `unpack()` inside `resolve_cell`?

**Comment by laurmaedje** [comment]
> It worked and was fairly simple, but it did introduce some perhaps unnecessary coupling with Packed, so I set off to change that.

Everything is coupled with `Packed`. That's okay.

> However, it did require one or two workarounds due to borrow checking stuff (in particular for the self.push_align(...) stuff since self.align() was being used, leading to concurrent mut / non-mut borrows due to the implicit deref/deref_mut), but right now I think I prefer this approach so as to keep CellGrid::resolve as generic and simplified as possible.

Yeah, that's also a thing in a few other parts of the codebase now.

> If we keep my current approach of implementing ResolvableCell on Packed<GridCell/TableCell> (at https://github.com/typst/typst/commit/4b37f741a05b3c6490c4ca96b1cb1cdf55248f45), is it fine to mutate the Packed directly like I'm doing, or should I always unpack() inside resolve_cell?

Mutating the packed is fine. Unpacking is actually worse because it'd lose the spans and other metadata that's stored in the `Packed`.

TLDR: Everything looks good as-is.

**Comment by laurmaedje** [comment]
I'm excited for advanced tables slowly coming together. 🎉

**Comment by PgBiel** [comment]
Same here :rocket:
Thanks!
